### ISSUE TITLE
[BUG] processing.AverageFramesByStep / piezo_movement_correction.correct_target_positions wrapping issue

### ISSUE THREAD

**Issue by barentine** [python-microscopy/python-microscopy#575]
**Describe the bug**
should only have one frame per z at the end, but looks like we wrap on the end. Will need to look into whether we are effectively 1-indexing with starttime or something. Note that putting in a `frames` array into `correct_target_positions` which is too long will continue dumping the rest into the last z step. 

<details> <summary> printing z values and unique counts </summary>

```
DEBUG:PYME.recipes.processing:Averaged stack size: 865
DEBUG:PYME.recipes.processing:averaging by step
[48.757 48.775 48.801 48.802 48.803 48.804 48.815 48.828 48.829 48.831
 48.838 48.84  48.841 48.843 48.855 48.895 48.9   48.906 48.909 48.912
 48.913 48.919 48.92  48.921 48.922 48.935 48.961 48.975 48.989 48.991
 48.994 49.    49.001 49.015 49.055 49.062 49.063 49.064 49.066 49.068
 49.083 49.095 49.135 49.141 49.145 49.148 49.15  49.151 49.162 49.163
 49.175 49.202 49.215 49.226 49.229 49.23  49.231 49.232 49.238 49.24
 49.241 49.255 49.295 49.308 49.309 49.31  49.313 49.318 49.319 49.32
 49.323 49.335 49.375 49.388 49.389 49.392 49.401 49.402 49.403 49.415
 49.455 49.468 49.47  49.471 49.476 49.48  49.481 49.482 49.483 49.495
 49.535 49.548 49.549 49.551 49.559 49.561 49.562 49.575 49.615 49.62
 49.628 49.629 49.63  49.632 49.634 49.64  49.641 49.642 49.655 49.695
 49.704 49.709 49.71  49.711 49.717 49.72  49.721 49.722 49.735 49.775
 49.78  49.783 49.788 49.789 49.792 49.793 49.796 49.799 49.801 49.802
 49.815 49.855 49.866 49.869 49.87  49.871 49.874 49.881 49.882 49.895
 49.935 49.948 49.952 49.955 49.956 49.957 49.958 49.96  49.961 49.962
 49.963 49.975 50.001 50.015 50.023 50.024 50.025 50.026 50.039 50.042
 50.055 50.095 50.105 50.106 50.107 50.108 50.121 50.123 50.135 50.175
 50.189 50.191 50.193 50.198 50.2   50.202 50.203 50.215 50.255 50.267
 50.268 50.269 50.272 50.281 50.282 50.283 50.295 50.335 50.348 50.35
 50.352 50.359 50.36  50.361 50.362 50.363 50.375 50.395 50.415 50.425
 50.427 50.428 50.429 50.43  50.437 50.439 50.44  50.455 50.495 50.505
 50.507 50.508 50.509 50.51  50.514 50.521 50.522 50.523 50.535 50.575
 50.587 50.588 50.589 50.59  50.599 50.6   50.602 50.603 50.615 50.655
 50.66  50.665 50.666 50.667 50.669 50.67  50.675 50.68  50.681 50.682
 50.695 50.735 50.741 50.747 50.749 50.75  50.754 50.759 50.76  50.761
 50.763 50.775 50.815 50.827 50.828 50.829 50.832 50.833 50.837 50.84
 50.842 50.855 50.877 50.88  50.895 50.908 50.909 50.918 50.921 50.922
 50.935 50.975 50.989 50.994 50.997 50.999 51.    51.001 51.002 51.015
 51.036 51.055 51.062 51.065 51.066 51.081 51.083 51.095 51.116 51.135
 51.148 51.152 51.158 51.162 51.163 51.175 51.215 51.228 51.23  51.232
 51.234 51.24  51.242 51.243 51.244 51.255 51.295 51.309 51.311 51.315
 51.319 51.32  51.321 51.323 51.335 51.375 51.388 51.389 51.39  51.391
 51.398 51.401 51.402 51.403 51.415 51.441 51.455 51.467 51.47  51.48
 51.481 51.482 51.495 51.517 51.535 51.547 51.55  51.551 51.553 51.559
 51.56  51.561 51.563 51.575 51.615 51.625 51.631 51.632 51.635 51.638
 51.639 51.64  51.641 51.642 51.655 51.68  51.695 51.709 51.71  51.712
 51.718 51.719 51.721 51.722 51.735 51.775 51.788 51.789 51.792 51.793
 51.8   51.804 51.815 51.855 51.869 51.87  51.873 51.88  51.881 51.882
 51.883 51.895 51.935 51.95  51.952 51.953 51.955 51.958 51.96  51.963
 51.975 52.015 52.023 52.025 52.028 52.029 52.04  52.041 52.042 52.043
 52.055 52.082 52.095 52.101 52.104 52.105 52.108 52.109 52.116 52.119
 52.123 52.135 52.175 52.189 52.19  52.191 52.199 52.2   52.202 52.215
 52.255 52.269 52.27  52.271 52.273 52.278 52.28  52.283 52.284 52.295
 52.335 52.344 52.35  52.351 52.352 52.354 52.359 52.36  52.364 52.375
 52.415 52.428 52.43  52.431 52.432 52.436 52.439 52.442 52.444 52.455
 52.495 52.511 52.512 52.513 52.519 52.52  52.521 52.522 52.535 52.575
 52.58  52.59  52.591 52.597 52.6   52.602 52.603 52.615 52.655 52.669
 52.67  52.671 52.672 52.673 52.678 52.681 52.682 52.695 52.717 52.735
 52.747 52.748 52.75  52.751 52.754 52.758 52.761 52.762 52.775 52.815
 52.829 52.83  52.831 52.836 52.84  52.841 52.855 52.895 52.909 52.912
 52.916 52.921 52.922 52.923 52.935 52.975 52.984 52.991 52.992 52.993
 53.001 53.002 53.003 53.015 53.055 53.062 53.064 53.066 53.067 53.068
 53.08  53.081 53.082 53.095 53.135 53.142 53.143 53.149 53.15  53.151
 53.153 53.16  53.162 53.175 53.215 53.223 53.228 53.231 53.233 53.234
 53.237 53.241 53.242 53.243 53.255 53.295 53.309 53.311 53.312 53.318
 53.322 53.324 53.335 53.364 53.375 53.388 53.39  53.393 53.4   53.402
 53.403 53.404 53.415 53.421 53.455 53.469 53.47  53.472 53.476 53.481
 53.482 53.483 53.495 53.535 53.548 53.549 53.551 53.553 53.561 53.562
 53.575 53.615 53.626 53.629 53.632 53.633 53.634 53.639 53.64  53.642
 53.655 53.695 53.711 53.712 53.713 53.719 53.72  53.722 53.723 53.735
 53.775 53.785 53.791 53.792 53.794 53.799 53.801 53.802 53.803 53.815
 53.855 53.869 53.87  53.872 53.881 53.882 53.895 53.919 53.935 53.948
 53.949 53.95  53.959 53.96  53.962 53.963 53.975 54.015 54.02  54.024
 54.027 54.032 54.037 54.04  54.041 54.042 54.055 54.095 54.101 54.106
 54.108 54.121 54.122 54.123 54.135 54.16  54.175 54.187 54.19  54.191
 54.192 54.201 54.202 54.215 54.255 54.269 54.27  54.271 54.272 54.279
 54.281 54.283 54.295 54.335 54.348 54.351 54.352 54.355 54.357 54.358
 54.361 54.362 54.375 54.415 54.423 54.429 54.43  54.431 54.432 54.441
 54.442 54.443 54.455 54.495 54.51  54.513 54.517 54.518 54.521 54.522
 54.535 54.575 54.581 54.589 54.59  54.592 54.598 54.599 54.6   54.615
 54.655 54.669 54.672 54.68  54.681 54.682 54.684 54.695 54.735 54.74
 54.75  54.751 54.757 54.758 54.759 54.762 54.775 54.799 54.815 54.829
 54.831 54.833 54.837 54.839 54.841 54.855 54.873 54.895 54.903 54.91
 54.911 54.914 54.919 54.922 54.935 54.975 54.981 54.991 54.992 54.993
 54.995 54.996 54.997 55.    55.001 55.002 55.015 55.055 55.067 55.068
 55.069 55.079 55.081 55.082 55.095 55.135 55.148 55.149 55.152 55.16
 55.162 55.163 55.175 55.215 55.221 55.223 55.229 55.23  55.231 55.232
 55.233 55.242 55.243 55.255 55.295 55.311 55.312 55.313 55.319 55.32
 55.322 55.335 55.375 55.381 55.391 55.392 55.398 55.401 55.402 55.403
 55.415 55.455 55.467 55.469 55.471 55.472 55.474 55.48  55.481 55.482
 55.483 55.495 55.516 55.535 55.549 55.552 55.553 55.559 55.56  55.562
 55.563 55.575 55.604 55.615 55.625 55.632 55.633 55.635 55.636 55.637
 55.638 55.641 55.642 55.655 55.695 55.709 55.711 55.713 55.714 55.718
 55.719 55.721 55.723 55.735 55.775 55.791 55.792 55.793 55.794 55.798
 55.799 55.8   55.801 55.815 55.855 55.87  55.872 55.873 55.874 55.879
 55.88  55.882 55.895 55.935 55.951 55.952 55.953 55.957 55.959 55.963
 55.975 55.998 56.001 56.002]
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 2]

```
</details>

**Comment by barentine** [comment]
Issue could be related to using time.time() as start time on line 1765 of `PYME.recipes.processing`, which might be revealing something similar to #576 as far as float comparisons go

<details>

setup so to speak
```
import numpy as np
from PYME.IO.MetaDataHandler import NestedClassMDHandler

# basePiezo position - offset = OffsetPiezo position
# Note that the ProtocolFocus events know about the offset (ie are logged by
# the OffsetPiezo), while the PiezoOnTarget events are logged by the BasePiezo.

TEST_EVENTS = np.array([('PiezoOnTarget', 0, '48.262'),
       ('ProtocolFocus', 0, '0, 49.988'),
       ('PiezoOnTarget', 1 * 0.00125, '48.307'),
       ('ProtocolTask', 1 * 0.00125, '1, DisableLock, '),
       ('PiezoOffsetUpdate', 2 * 0.00125, '-1.6720'),
       ('ProtocolFocus', 801 * 0.00125, '801, 51.188'),
       ('PiezoOnTarget', 850 * 0.00125, '49.489'),
       ('ProtocolFocus', 1601 * 0.00125, '1601, 52.388'),
       ('PiezoOnTarget', 1650 * 0.00125, '50.705'),
       ('ProtocolFocus', 2401 * 0.00125, '2401, 53.588'),
       ('ProtocolTask', 2501 * 0.00125, 'EnableLock, '),
       ('ProtocolTask', 2501 * 0.00125, 'LaunchAnalysis, ')],
                       # fixme - but the S32 and S256 back to unicode once we fix event typing elsewhere
      dtype=[('EventName', 'S32'), ('Time', '<f8'), ('EventDescr', 'S256')])

CHANGES = np.array([(0, 0, 48.262, 48.262),
       (1 * 0.00125, 1, 48.307, 48.307),
       (801 * 0.00125, 801, 51.188, 51.188),
       (850 * 0.00125, 850, 49.489 + 1.6720, 49.489),
       (1601 * 0.00125, 1601, 52.388, 52.388),
       (1650 * 0.00125, 1650, 50.705 + 1.6720, 50.705),
       (2401 * 0.00125, 2401, 53.588, 53.588)],
       dtype= [('t', '<f8'), ('frame', '<i4'), ('z', '<f8'), 
               ('z_ignoring_offset', '<f8')])

TEST_DATA_SOURCE = np.arange(2500).astype([('t', '<i4')])
GROUND_TRUTH_Z = np.empty(len(TEST_DATA_SOURCE), dtype=float)
GROUND_TRUTH_Z_IGNORING_OFFSET = np.empty_like(GROUND_TRUTH_Z)
I = np.argsort(CHANGES['t'])
CHANGES = CHANGES[I]
for frame, z, z_no_off in zip(CHANGES['frame'], CHANGES['z'], CHANGES['z_ignoring_offset']):
    GROUND_TRUTH_Z[frame:] = z
    GROUND_TRUTH_Z_IGNORING_OFFSET[frame:] = z_no_off


TEST_MDH = NestedClassMDHandler()
TEST_MDH['Camera.CycleTime'] = 0.00125
TEST_MDH['StartTime'] = 0
```

and some testing

```
In [22]:     from PYME.Analysis.piezo_movement_correction import correct_target_positions
    ...: 
    ...:     corrected_focus = correct_target_positions(TEST_DATA_SOURCE['t'], TEST_EVENTS, TEST_MDH)
    ...:     np.testing.assert_array_almost_equal(GROUND_TRUTH_Z, corrected_focus)
    ...: 
    ...:     # now fake the same thing from events
    ...:     frames = np.arange(len(GROUND_TRUTH_Z))
    ...:     # t = TEST_MDH['StartTime']
    ...:     t = time.time()
    ...:     cycle_time = TEST_MDH['Camera.CycleTime']
    ...:     fudged_events = []
    ...:     step_inds = np.arange(len(CHANGES), dtype=int)
    ...:     for si in step_inds:
    ...:         fudged_events.append(('ProtocolFocus', t, '%d, %3.3f' % (si, CHANGES[si]['z'])))
    ...:         fudged_events.append((('StartAq', t, '%d' % si)))
    ...:         t += cycle_time
    ...: 
    ...:     fudged_events = np.array(fudged_events, dtype=[('EventName', 'S32'),
    ...:                                                    ('Time', '<f8'),
    ...:                                                    ('EventDescr', 'S256')])
    ...:     z_vals = correct_target_positions(step_inds, fudged_events, TEST_MDH)
    ...:     uni, counts = np.unique(z_vals, return_counts=True)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-22-8639ab074ba2> in <module>
      7 frames = np.arange(len(GROUND_TRUTH_Z))
      8 # t = TEST_MDH['StartTime']
----> 9 t = time.time()
     10 cycle_time = TEST_MDH['Camera.CycleTime']
     11 fudged_events = []

NameError: name 'time' is not defined

In [23]: import time

In [24]:     from PYME.Analysis.piezo_movement_correction import correct_target_positions
    ...: 
    ...:     corrected_focus = correct_target_positions(TEST_DATA_SOURCE['t'], TEST_EVENTS, TEST_MDH)
    ...:     np.testing.assert_array_almost_equal(GROUND_TRUTH_Z, corrected_focus)
    ...: 
    ...:     # now fake the same thing from events
    ...:     frames = np.arange(len(GROUND_TRUTH_Z))
    ...:     # t = TEST_MDH['StartTime']
    ...:     t = time.time()
    ...:     cycle_time = TEST_MDH['Camera.CycleTime']
    ...:     fudged_events = []
    ...:     step_inds = np.arange(len(CHANGES), dtype=int)
    ...:     for si in step_inds:
    ...:         fudged_events.append(('ProtocolFocus', t, '%d, %3.3f' % (si, CHANGES[si]['z'])))
    ...:         fudged_events.append((('StartAq', t, '%d' % si)))
    ...:         t += cycle_time
    ...: 
    ...:     fudged_events = np.array(fudged_events, dtype=[('EventName', 'S32'),
    ...:                                                    ('Time', '<f8'),
    ...:                                                    ('EventDescr', 'S256')])
    ...:     z_vals = correct_target_positions(step_inds, fudged_events, TEST_MDH)
    ...:     uni, counts = np.unique(z_vals, return_counts=True)

In [25]: uni
Out[25]: array([48.262, 48.307, 51.161, 51.188, 52.377, 52.388])

In [26]: counts
Out[26]: array([1, 1, 1, 1, 2, 1])
```

</details>

**Comment by barentine** [comment]
or using time.time() instead of mdh['StartTime']

**Comment by barentine** [comment]
yep, that was it

**Comment by barentine** [comment]
fixed in #576
